### PR TITLE
Keycloak 12624 nullpointer during OIDC logout client disabled

### LINKED ISSUE #12624: Possible NullPointerException during OpenID Connect logout with disabled client
### Describe the bug

During migration from Keycloak 9.0.3 to 18.0.0 (Quarkus), we noticed a `NullPointerException` during OpenID Connect logout.

Some of our internal applications still use the legacy logout parameter `redirect_uri`. Based on https://www.keycloak.org/2022/04/keycloak-1800-released#_openid_connect_logout, we start Keycloak 18.0.0 with parameter `spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true`: `bin/kc.[sh|bat] --spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true start`

### Version

18.0.0

### Expected behavior

_No response_

### Actual behavior

* `NullPointerException` after confirmation of Keycloak OpenID Connect logout

```
ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-97) Uncaught server error: java.lang.NullPointerException
        at org.keycloak.services.resources.SessionCodeChecks.initialVerify(SessionCodeChecks.java:228)
        at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint.logoutConfirmAction(LogoutEndpoint.java:262)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:201)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:201)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
        at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

### How to Reproduce?

* Open OpenID Connect logout URL with parameter `redirect_url`
* Keycloak shows logout confirm dialog
* Confirm logout

### Anything else?

* https://www.keycloak.org/2022/04/keycloak-1800-released#_openid_connect_logout

### PR DESCRIPTION
**Pull request by marcelomrwin**
Fixes #12624

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java [hunk 1]
```diff
@@ -222,9 +224,9 @@ public boolean initialVerify() {
         setClientToEvent(client);
         session.getContext().setClient(client);
 
-        if (!client.isEnabled()) {
+        if (!client.isEnabled() && clientCode != null) {
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:227]: One note: Is it possible to make sure that `SessionCodeChecks` is "effectively unchanged" ?

The thing is, that `SessionCodeChecks` is used also during authentication and I fear of the possible security issues. I am thinking about something like:
- add the protected method like `checkClientDisabled` in the `SessionCodeChecks` class. This protected method will do exactly same thing as today in the `SessionCodeChecks` class. Something like:
```
protected boolean checkClientDisabled() {
    return !client.isEnabled;
}
- In the `LogoutSessionCodeChecks`, this protected method can be overriden and check changed to something like:
```
protected boolean checkClientEnabled() {
    return !client.isEnabled && clientCode != null;
}
```
Will this work?

**Review by mposolda — CHANGES_REQUESTED** [review]
@marcelomrwin Thanks for the PR!

I've added one inline comment (Sorry if I was unclear when we had a chat about this today morning. Hope it is clearer from the comment)

Besides that, is it possible to write automted test also for the scenario when the `account` client is disabled? I am thinking about the scenario like:
- User login
- Client `account` will be disabled
- Logout is triggered by going to the LogoutEndpoint without any parameters (In case that none of `id_token_hint` and `client_id` parameters is used, then the "System client" is used as the client during logout. And this system client is the `account` client). See calls to `SystemClientUtil.getSystemClient` for more details

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 1 above)
Comment (marcelomrwin) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:227]: I understand. A strategy model to ensure that according to the operation we will carry out the necessary verification. Implemented as per your guidance.

**Comment by marcelomrwin** [comment]
> @marcelomrwin Thanks for the PR!
> 
> I've added one inline comment (Sorry if I was unclear when we had a chat about this today morning. Hope it is clearer from the comment)
> 
> Besides that, is it possible to write automted test also for the scenario when the `account` client is disabled? I am thinking about the scenario like:
> 
> * User login
> * Client `account` will be disabled
> * Logout is triggered by going to the LogoutEndpoint without any parameters (In case that none of `id_token_hint` and `client_id` parameters is used, then the "System client" is used as the client during logout. And this system client is the `account` client). See calls to `SystemClientUtil.getSystemClient` for more details

Test scenario created and tested. Working as expected. I await your appreciation to ensure that the understanding was correct.

**Review by mposolda — APPROVED** [review]
@marcelomrwin Thanks Marcelo for the updates and for the automated test!

**Pull request merged**